### PR TITLE
Allow iterating ctx.secrets.items()

### PR DESCRIPTION
**Pull request by kaixi-wang**
## What changes are proposed in this pull request?

Re-enable iterating through secrets like a dict (TEAMS-1980)

### CODE REVIEW

File: fiftyone/plugins/secrets.py [hunk 1]
```diff
@@ -118,6 +125,12 @@ def __init__(self, secrets_dict, operator_uri=None, resolver_fn=None):
     def __eq__(self, other):
         return self.__secrets == other
 
+    def __setattr__(self, key, value):
+        if key == "__secrets" or key == "__required_keys":
```
Comment (brimoor) [inline fiftyone/plugins/secrets.py:129]: This isn't necessary. Double underscore attributes already aren't accessible via `d.__secrets` from outside the class. Python does weird name mangling that makes them available as `d._SecretsDictionary__secrets`.

Either way, don't waste effort trying to prevent things from happening if users modify private variables. It's Python. If the user really wants to modify private variables then they can mess things up infinitely worse than just overriding secrets.

File: fiftyone/plugins/secrets.py [hunk 2]
```diff
@@ -155,5 +168,7 @@ def values(self):
         raise RuntimeError("Iteration through values is not allowed")
 
     def items(self):
-        # Override items() to prevent iterating through plaintext values
-        raise RuntimeError("Iteration through items is not allowed.")
+        # Override items() to use __getitem__ to automatically resolve
```
Comment (brimoor) [inline fiftyone/plugins/secrets.py:171]: `values()` should be allowed too. It's a documented as a dict, so users should be allowed to do dict-like things

File: fiftyone/plugins/secrets.py [hunk 3]
```diff
@@ -155,5 +168,7 @@ def values(self):
...
+        # missing secrets upon iteration
+        for key in self.keys():
+            yield key, self[key]
```
Comment (brimoor) [inline fiftyone/plugins/secrets.py:174]: IDK how the resolver is implemented in teams, but need to verify: is calling `self[key]` in a loop performant? IE will it only make one API call to pull in all secrets?

There are plugins that declare 10+ secrets, so 10 API requests would be 😢

**Review by brimoor — CHANGES_REQUESTED** [review]
Inheriting from `Mapping` would give you a clean way to implement a read-only dict interface:

```py
from collections.abc import Mapping

class SecretsDictionary(Mapping):

    def __init__(self, secrets):
        self._secrets = secrets

    def __getitem__(self, key):
        # resolve here
        return self._secrets[key]

    def __len__(self):
        return len(self._secrets)

    def __iter__(self):
        return iter(self._secrets)

d = SecretsDictionary({"foo": "bar", "spam": "eggs"})
```

```py
print(d["foo"])  # bar
print(d["spam"])  # eggs
print(len(d))  # 2
print(list(d))  # ['foo', 'spam']
print(list(d.keys()))  # ['foo', 'spam']
print(list(d.values()))  # ['bar', 'eggs']
print(list(d.items()))  # [('foo', 'bar'), ('spam', 'eggs')]

d["foo"] = "hi"  # TypeError: 'SecretsDictionary' object does not support item assignment
```

File: fiftyone/plugins/secrets.py (hunk 3 above)
Comment (kaixi-wang) [reply fiftyone/plugins/secrets.py:174]: this is the best option right now. Anyways, the secrets are only potentially unresolved in resolve_input(). Any usage of the secrets in execute or resolve_ouput() should already have batch resolved the secrets. Moreover, self[key] will not only resolve it, but add it to the secrets dict so the maximum number of requests is the number of unique secrets. If performance is an issue, then it is likely the number of times the context is regenerated/operators are initialized.

File: fiftyone/plugins/secrets.py (hunk 2 above)
Comment (kaixi-wang) [reply fiftyone/plugins/secrets.py:171]: I'm against this, but fine

File: fiftyone/plugins/secrets.py (hunk 1 above)
Comment (kaixi-wang) [reply fiftyone/plugins/secrets.py:129]: I don't see it as wasted effort because at the end of the day if they mess it up, it comes back to me to debug. I'd rather prevent mutation from the start rather than allow it.

**Review by ritch — APPROVED** [review]

**Comment by ritch** [comment]
I've tested this locally and can mutate the secrets and required_keys.

```
d = SecretsDictionary({"foo": "bar"}, required_keys=["foo"])
d._SecretsDictionary__required_keys.append('bar')
d._SecretsDictionary__secrets = {'bar': 'bat'}

d.keys()
# => ['bar']
```

**Review by ritch — CHANGES_REQUESTED** [review]
Please fix the ability to mutate `__secrets` and `__required_keys`.

**Comment by kaixi-wang** [comment]
> Please fix the ability to mutate `__secrets` and `__required_keys`.

@brimoor requested I remove that part...

**Comment by kaixi-wang** [comment]
Disabled setting the value. Attempts will throw an error 
```
Error loading operators from server

Traceback (most recent call last):
     ...
    fiftyone/plugins/secrets.py", line 223, in __setattr__
    raise KeyError("Cannot mutate hidden properties")
KeyError: 'Cannot mutate hidden properties'
```

File: fiftyone/plugins/secrets.py [hunk 4]
```diff
@@ -5,8 +5,11 @@
 | `voxel51.com <https://voxel51.com/>`_
 |
 """
+# from __future__ import annotations
```
Comment (brimoor) [inline fiftyone/plugins/secrets.py:8]: please remove commented code

File: fiftyone/plugins/secrets.py [hunk 5]
```diff
@@ -5,8 +5,11 @@
...
 import logging
+import re
 import typing
+from collections.abc import Mapping
```
Comment (brimoor) [inline fiftyone/plugins/secrets.py:12]: nit: style guide says to alphabetize imports

File: fiftyone/plugins/secrets.py [hunk 6]
```diff
@@ -101,20 +104,35 @@ def _get_secrets_client():
     return client()
 
 
-class SecretsDictionary:
+class SecretsDictionary(Mapping[str, str]):
```
Comment (brimoor) [inline fiftyone/plugins/secrets.py:107]: just checking: is this syntax valid in Python 3.7? We still promise support for it

File: fiftyone/plugins/secrets.py [hunk 7]
```diff
@@ -101,20 +104,35 @@ def _get_secrets_client():
...
         if resolver_fn:
             self._resolver = resolver_fn
         else:
             self._resolver = None
 
+    def __len__(self):
+        """Returns the number of secrets defined in the plugin definition"""
+        return len(self.__required_keys)
+
+    def __iter__(self):
```
Comment (brimoor) [inline fiftyone/plugins/secrets.py:132]: For dicts, `__iter__` should only return keys, not values

File: fiftyone/plugins/secrets.py [hunk 8]
```diff
@@ -128,32 +146,36 @@ def __getitem__(self, key):
...
+        logging.warning("Copying the SecretsDictionary values is not allowed.")
         return self.__deepcopy__()
 
     def keys(self):
         return [k for k, v in self.__secrets.items() if v is not None]
 
     def values(self):
-        # Override values() to prevent iterating through plaintext values
-        raise RuntimeError("Iteration through values is not allowed")
+        # Override values() to ensure that resolvable secrets are always returned upon iteration
```
Comment (brimoor) [inline fiftyone/plugins/secrets.py:174]: My suggestion to inherit from `Mapping` had two benefits:
- it automatically handles throwing errors when mutation methods like `__setitem__` and `update()` are called
- you only have to implement `__getitem__`, `__len__`, and `__iter__` as described [here](https://github.com/voxel51/fiftyone/pull/3820#pullrequestreview-1740197017) and then other methods like `keys()`, `values()`, and `items()` will automatically use those in the correct ways.

In other words, once you fix your `__iter__` implementation I believe `keys()`, `values()`, and `items()` can just be deleted.

File: fiftyone/plugins/secrets.py [hunk 9]
```diff
@@ -128,32 +146,36 @@ def __getitem__(self, key):
                 self.__secrets[key] = val
         return val
 
+    def __setattr__(self, key, value):
+        if re.search("_SecretsDictionary__(secrets|required_keys)$", key):
```
Comment (brimoor) [inline fiftyone/plugins/secrets.py:150]: Still don't see a point in preventing `__setattr__` here when the user can still do other things like `d._SecretsDictionary__secrets["bad"] = "thing"`.

File: fiftyone/plugins/secrets.py [hunk 10]
```diff
@@ -128,32 +146,36 @@ def __getitem__(self, key):
...
+            super().__setattr__(key, value)
+
     def __setitem__(self, key, value):
         raise RuntimeError("Setting values is not allowed")
 
     def __deepcopy__(self, memodict={}):
-        logging.warning(
-            "Copying the SecretDictionary values is not " "allowed."
-        )
+        logging.warning("Copying the SecretsDictionary values is not allowed.")
```
Comment (brimoor) [inline fiftyone/plugins/secrets.py:159]: Why log a warning that something is now allowed but let it happen anyway?

Also why is this method implemented? It returns a regular dict so is not actually copying the object, it returns a different type of object.

Why not let the default implementations of `copy()` and `__deepcopy__` do their things?

**Review by brimoor — APPROVED** [review]
Approving to unblock.

The `__iter__` implementation change needs to happen, and I don't understand why `__deepcopy__` is implemented.

Other comments would cleanup code.

File: fiftyone/plugins/secrets.py (hunk 9 above)
Comment (kaixi-wang) [reply fiftyone/plugins/secrets.py]: currently, there will be no error if you manually set the secret, however, it will not show up when you iterate through ctx.secrets.items()

File: fiftyone/plugins/secrets.py (hunk 6 above)
Comment (kaixi-wang) [reply fiftyone/plugins/secrets.py]: removed

File: fiftyone/plugins/secrets.py (hunk 5 above)
Comment (kaixi-wang) [reply fiftyone/plugins/secrets.py]: removed

File: fiftyone/plugins/secrets.py (hunk 7 above)
Comment (brimoor) [reply fiftyone/plugins/secrets.py:140]: Fixing this is important because users often iterate over dicts like this:

```py
for key in ctx.secrets:
    value = ctx.secrets[key]
```

File: fiftyone/plugins/secrets.py (hunk 10 above)
Comment (kaixi-wang) [reply fiftyone/plugins/secrets.py:197]: To discourage copying the secret values and passing them around and encourage usage exclusively via ctx.secret/s both for security reasons and to avoid potential issues with copying the value to an object and expecting that the object have the secrets when called in another part of the code due to the execution context potentially being reset inbetween

File: fiftyone/plugins/secrets.py (hunk 7 above)
Comment (kaixi-wang) [reply fiftyone/plugins/secrets.py:140]: fixed

File: fiftyone/plugins/secrets.py (hunk 9 above)
Comment (ritch) [reply fiftyone/plugins/secrets.py]: Would overriding `__getattr__` be the solution to reading/mutating the underlying dictionary?

**Review by ritch — APPROVED** [review]
LGTM
